#46-Update-Controller

mliam0608 opened the pull request:
Added a signalsController.cpp and signalsController.h file. The new signalsController class can check and change the status for different signals

## Changes

Added a SignalsController class inside the controllers folder.

## Notes

I have not tested the code yet.

Closes #46

mliam0608 committed:
#46-Update-Controller

Added a signalsController.cpp and signalsController.h file. The new signalsController class can check and change the status for different signals

Peyton-McKee commented on NERODevelopment/src/controllers/signalsController.h:
```diff
@@ -0,0 +1,29 @@
+ * @brief The TheSignalsController class
+ * Controller for the signals view
+ */
+class SignalsController : public QObject {
```
Rename to OffViewController

mliam0608 replied:
Done!

Peyton-McKee replied:
Still this, you are creating one controller

Peyton-McKee commented on NERODevelopment/src/controllers/signalscontroler.cpp:
```diff
@@ -0,0 +1,33 @@
+{
+    // Inital mockup of statuses, Strings represent
+    // the name and status of each attribute
+    attributeStatus["SideBRBs"] = "unchecked";
```
can you make these macros

mliam0608 replied:
I added these in AttributeStatus.h

Peyton-McKee replied:
you didnt actually use the macros?

Peyton-McKee commented on NERODevelopment/src/controllers/signalsController.h:
```diff
@@ -0,0 +1,29 @@
+    Q_OBJECT
+    Q_PROPERTY(QMap<QString, QString> attributeStatus READ getAttributeStatus NOTIFY attributeStatusChanged)
+public:
+    explicit SignalsController(QObject *parent = nullptr);
```
you also need properties for pack temp, motor temp and state of charge

mliam0608 replied:
I added pointers to reference these properties from the other files. Does that work?

Peyton-McKee replied:
where

Peyton-McKee commented on NERODevelopment/src/controllers/signalscontroler.cpp:
```diff
@@ -0,0 +1,33 @@
+    attributeStatus["BSPD"] = "unchecked";
+    attributeStatus["MPU"] = "unchecked";
+    attributeStatus["BOTS"] = "unchecked";
+    attributeStatus["INERTIA"] = "unchecked";
```
also the unchecked should be an enumeration of possible values, so like good, faulted, off etc

Peyton-McKee replied:
you can do that through macros in a file

mliam0608 replied:
Done!

Peyton-McKee commented on NERODevelopment/src/controllers/signalscontroler.cpp:
```diff
@@ -0,0 +1,33 @@
+    return attributeStatus.value(attribute);
+}
+
+void SignalsController::setAttributeStatus(const QString &attribute, const QString &status)
```
you need to make an update function that calls the models get functions for all of these values and sets the attribute status for all of them. Your probably going to need to make the model functions for that as well

Peyton-McKee reviewed (changes requested):
Good start, Make sure to rename the file to OffViewController

mliam0608 committed:
Added names and status headers, with model class to combine them.

I added an AttributeNames header o define possible names for signals as macros. I also added an AttributeStatus header file to define enumeration values for possible statuses. To create a dictionary value for each signal, I created an offViewModel class that assigns a status to a name.

Peyton-McKee commented on NERODevelopment/src/controllers/offViewController.cpp:
```diff
@@ -0,0 +1,28 @@
+#include "offViewController.h"
```
this file doesnt follow current naming conventions

Peyton-McKee commented on NERODevelopment/src/controllers/attributeNames.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef ATTRIBUTENAMES_H
```
file doesnt follow naming conventions

Peyton-McKee commented on NERODevelopment/src/controllers/attributeNames.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef ATTRIBUTENAMES_H
+#define ATTRIBUTENAMES_H
+
+#define SIDEBRBS "SideBRBs"
```
All caps

Peyton-McKee commented on NERODevelopment/src/controllers/attributestatus.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef ATTRIBUTESTATUS_H
+#define ATTRIBUTESTATUS_H
+
+enum class AttributeStatus {
```
assign raw values to them, like = 0, 1, 2

Peyton-McKee commented on NERODevelopment/src/controllers/offviewmodel.cpp:
```diff
@@ -0,0 +1,13 @@
+#include "OffViewModel.h"
+
+OffViewModel::OffViewModel(QObject *parent) : QObject(parent)
```
no this file shouldnt exist, you should add getters to the current model for all of these values and add them to the current macro file for the model

Peyton-McKee commented on NERODevelopment/src/controllers/offviewmodel.h:
```diff
@@ -0,0 +1,18 @@
+#ifndef OFFVIEWMODEL_H
+#define OFFVIEWMODEL_H
```
same with this

Peyton-McKee commented on NERODevelopment/src/controllers/offviewmodel.cpp:
```diff
@@ -0,0 +1,13 @@
+    }
+}
+
+AttributeStatus OffViewModel::getAttributeStatus(AttributeName attributeName) const
```
if anything make this a function in the actual model

Peyton-McKee reviewed (changes requested):
You should be only making one controller. You need to delete the signals controller if your not using it anymore, also you still need to add pack, motor, state of charge as other q properties. Make sure to tag your commits with the issue number, and also dont make a new model, just ad the function to the current models

mliam0608 committed:
Delete NERODevelopment/src/controllers/signalsController.h

mliam0608 committed:
Delete NERODevelopment/src/controllers/signalscontroler.cpp

mliam0608 committed:
#46 Delete NERODevelopment/src/controllers/offviewmodel.cpp

mliam0608 committed:
#46 Delete NERODevelopment/src/controllers/offviewmodel.h

mliam0608 committed:
#53 Update-Controller

Changed attributeStatus enumerations to raw values. Added Q_Properties for pack temp, motor temp, and state of charge

mliam0608 committed:
Merge branch '#46-Update-Controller' of https://github.com/Northeastern-Electric-Racing/Nero-2.0 into #46-Update-Controller

mliam0608 committed:
#46 Rename attributeNames.h to attributenames.h

mliam0608 committed:
#46-Update-Controller

Update attributeStatus enumerations and add packTemp, stateOfCharge, and motorTemp QProperties to offViewController.

mliam0608 committed:
#46-Update-Controller

Remove unnecessary efficiencyController.h and chargingController.h imports.

Peyton-McKee commented on NERODevelopment/src/controllers/offViewController.cpp:
```diff
@@ -0,0 +1,54 @@
+#include "offviewcontroller.h"
```
follow file naming conventions

Peyton-McKee commented on NERODevelopment/src/controllers/offViewController.h:
```diff
@@ -0,0 +1,49 @@
+
+    QMap<QString, AttributeStatus> getAttributeStatus() const;
+    int packTemp() const;
+    qreal motorTemp() const;
```
we use floats i think

Peyton-McKee commented on NERODevelopment/src/controllers/offViewController.cpp:
```diff
@@ -0,0 +1,54 @@
+
+void OffViewController::updateAttributesFromModel() {
+    for (const auto& attributeName : {SIDEBRBS, BMS, BSPD, MPU, BOTS, INERTIA, CPBRB, TSMS, HVDINTRLK, HVCNCTR}) {
+        setAttributeStatus(attributeName, m_attributeStatus[attributeName]);
```
this is just going to set it to itself isnt it? You need to get the updated value from the model. You can use the getById Function I believe. Actually you can add the attribute names macros to the data_type_names.h macro file

mliam0608 committed:
#46-Update-controller

Change packTemp, stateOfCharge, and motorTemp to floats. Introduce a model field for the offViewController class.

mliam0608 committed:
Rename offViewController.cpp to offviewcontroller.cpp

mliam0608 committed:
Rename offViewController.h to offviewcontroller.h

mliam0608 committed:
#46-Update-controller

Add model import line. Change offViewController constructor to initialize packTemp, motorTemp, and stateOfCharge values. Change updateAttributesFromModel function to properly update attribute statuses.

Peyton-McKee reviewed (changes requested):
you need to connect the model and the controller, and delete the viewmodel attribute, also move the macros to the data_type_name file

mliam0608 committed:
#46-update-controller

Connect model in constructor.

mliam0608 committed:
Merge branch '#46-Update-Controller' of https://github.com/Northeastern-Electric-Racing/Nero-2.0 into #46-Update-Controller

mliam0608 committed:
#46-Update-controller

Move attributeNames to data_type_names file.

mliam0608 committed:
#46-Update-controller

Delete attributenames.h file.

mliam0608 committed:
Merge branch '#46-Update-Controller' of https://github.com/Northeastern-Electric-Racing/Nero-2.0 into #46-Update-Controller

Peyton-McKee commented on NERODevelopment/src/controllers/offviewcontroller.cpp:
```diff
@@ -0,0 +1,50 @@
+
+void OffViewController::updateAttributesFromModel() {
+    const auto attributeStatusMap = m_model->getAttributeStatus();
+    for (const auto& attributeName : {SIDEBRBS, BMS, BSPD, MPU, BOTS, INERTIA, CPBRB, TSMS, HVDINTRLK, HVCNCTR}) {
```
you also need to update the pack temp, motor temp and state of charge, the model already has dedicated functions for these

Peyton-McKee reviewed (changes requested):
one last thing then you should be good

mliam0608 committed:
#46-Update-Controller

Added updates for packTemp, motorTemp, and stateOfCharge to updateAttributesFromModel() function.

Peyton-McKee commented on NERODevelopment/src/controllers/offviewcontroller.cpp:
```diff
@@ -0,0 +1,53 @@
+OffViewController::OffViewController(Model *model, QObject *parent)
+    : QObject{parent}, m_packTemp(0.0), m_motorTemp(0.0), m_stateOfCharge(0.0) {
+    this->m_model = model;
+    connect(m_model, &Model::updateAttributesFromModel, this,
```
this compiles? i dont think this function exists in the model, you want the models currentDataDidChange Signal

Peyton-McKee commented on NERODevelopment/src/controllers/attributestatus.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef ATTRIBUTESTATUS_H
```
move this file to utils

Peyton-McKee reviewed (changes requested):
You have to add these files to the actual project in the cmakelists.txt file under project sources. Then when you try to build youll see that it fails

mliam0608 committed:
#46-Update-controller

Move attributeStatus.h to utils folder, update offViewController constructor.

mliam0608 committed:
#46-Update-controller

Remove attributeStatus.h from the controllers folder.

Peyton-McKee reviewed (approved):
Beautiful
